### PR TITLE
Small bucket improvements

### PR DESCRIPTION
**Pull request by vogel**
# Description

This code adds asserts and changes Bucket to Bucket const to ensure that buckets are properly managed.

### CODE REVIEW

File: src/bucket/FutureBucket.cpp [hunk 1]
```diff
@@ -297,12 +329,15 @@ FutureBucket::startMerge(Application& app, bool keepDeadEntries)
...
 {
+    assertThreadIsMain();
+
     checkState();
     assert(!isLive());
     assert(hasHashes());
     auto& bm = app.getBucketManager();
     if (hasOutputHash())
     {
+        // THIS
```
Comment (marta-lokhova) [inline src/bucket/FutureBucket.cpp]: debugging leftover?

Comment (vogel) [reply src/bucket/FutureBucket.cpp]: oops

**Comment by MonsieurNicolas** [comment]
I would rather not refactor code like this. A good change started but not completed in this PR would be to move to using `Bucket const` in most places as per #1018 
Best to make this change would not be to add `const` everywhere though, but instead rename `Bucket` to something like `RawBucket` and define `Bucket` to be the `const` version of that type, that way, only a few places in the code base will have to be changed (only places are going to be where we actually create buckets).

**Comment by vogel** [comment]
using Bucket = RawBucket const; requires also multiple class RawBucket; before, so I don't think it is optimal solution.

I've also tried to go one step further, and make using Bucket = std::shared_ptr<RawBucket const>;, which I think is better solution, but still require multiple statements per file. As we have multiple ways of defining shared pointers now:

typedef std::shared_ptr<TxSetFrame> TxSetFramePtr;
class Config
{
  typedef std::shared_ptr<Config> pointer;
}

I think I'll do a step back and clean then up a bit (and make consistent). Maybe even some *_fwd.h includes will be needed, that will contain only typedefs?

**Comment by MonsieurNicolas** [comment]
sweeping stale pull requests

**Pull request closed without merging**